windows/code/DBStorage.h

```diff
@@ -1,71 +1,160 @@
-// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 #pragma once
 
+#include <optional>
+
 #include <winsqlite/winsqlite3.h>
 
 #include "NativeModules.h"
 
-class DBStorage
-{
-public:
-    typedef std::function<void(std::vector<winrt::Microsoft::ReactNative::JSValue> const &)>
-        Callback;
+struct DBStorage {
+    // To pass KeyValue pairs in the native module API.
+    // It has custom ReadValue and WriteValue to read/write to/from JSON.
+    struct KeyValue {
+        std::string Key;
+        std::string Value;
+    };
 
-    class DBTask
-    {
-    public:
-        enum class Type { multiGet, multiSet, multiRemove, clear, getAllKeys };
+    // An Error object for the native module API.
+    // It has a custom WriteValue to write to JSON.
+    struct Error {
+        std::string Message;
+    };
+
+    // An error list shared between Promise and DBTask.
+    struct ErrorHandler {
+        std::nullopt_t AddError(std::string &&message) noexcept;
+        const std::vector<Error> &GetErrors() const noexcept;
+
+    private:
+        std::vector<Error> m_errors;
+    };
+
+    // Ensure that only one result onResolve or onReject callback is called once.
+    template <typename TOnResolve, typename TOnReject>
+    struct Promise {
+
+        Promise(TOnResolve &&onResolve, TOnReject &&onReject) noexcept
+            : m_onResolve(std::move(onResolve)), m_onReject(std::move(onReject))
+        {
+        }
+
+        ~Promise()
+        {
+            Reject();
+        }
+
+        Promise(const Promise &other) = delete;
+        Promise &operator=(const Promise &other) = delete;
```

Comment:
Do we also need to delete the `operator=`s?

Reply:
Yes, to ensure that the type is not copyable we must delete the copy constructor and the copy assignment. We do not want allow this class to be copyable because it has an atomic variable that ensures that we call Resolve or Reject only once. If could copy the class state, we would lose such guarantee.

Reply:
For some reason, I thought we were _missing_ `operator=` here 😛  What about the move constructor/operator?

Reply:
The move constructor/operator= are removed automatically when we remove the copy constructor/operator=. If we want them enable, then they must be added back explicitly. They can be either `=default` or custom.